### PR TITLE
[tune] Add retry logic for restoring trials.

### PR DESCRIPTION
**Pull request by xwjiang2010**
Signed-off-by: xwjiang2010 <xwjiang2010@gmail.com>

## Why are these changes needed?

This is an advanced setting. Consider the following scenario: Due to scheduling glitches, sometimes a restoring
trial may be scheduled onto a dying node. By setting this env var to a positive number, the trial can be restored
several times and hopefully one of the times it will not be put on a dying node. This retry behavior won't increment
the per trial failure number, which is compared against `max_failures`.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by xwjiang2010** [comment]
@krfricke actually do you know why we don't physically delete those checkpoint_0000x folders when tune decides to restore from scratch? I think this a potential bug source.

File: doc/source/tune/api_docs/env.rst [hunk 1]
```diff
@@ -81,6 +81,13 @@ These are the environment variables Ray Tune currently considers:
   repeatedly every this amount of seconds. Defaults to 60 (seconds).
 * **TUNE_STATE_REFRESH_PERIOD**: Frequency of updating the resource tracking from Ray. Defaults to 10 (seconds).
 * **TUNE_SYNC_DISABLE_BOOTSTRAP**: Disable bootstrapping the autoscaler config for Docker syncing.
+* **TUNE_RESTORE_RETRY_NUM**: The number of retries that are done before a particular trial's restore is determined
+  unsuccessful. After that, the trial is not restored to its previous checkpoint but rather from scratch.
+  Default is 0.
+  This is an advanced setting. Consider the following scenario: Due to scheduling glitches, sometimes a restoring
+  trial may be scheduled onto a dying node. By setting this env var to a positive number, the trial can be restored
+  several times and hopefully one of the times it will not be put on a dying node. This retry behavior won't increment
+  the per trial failure number, which is compared against `max_failures`.
```
Comment (krfricke) [inline doc/source/tune/api_docs/env.rst:90]: This is a bit verbose, I think we can remove everything after `Default is 0`

File: python/ray/tune/execution/ray_trial_executor.py [hunk 2]
```diff
@@ -498,13 +499,21 @@ def _stop_trial(
...
         trial.set_location(_Location())
 
         try:
-            trial.write_error_log(exc=exc)
+            trial.write_error_log(
+                exc=exc,
+                should_increment_num_failures_counter=(
+                    should_increment_num_failures_counter
+                ),
+            )
```
Comment (krfricke) [inline python/ray/tune/execution/ray_trial_executor.py:516]: This leads to a pretty strong coupling of retry-logic between trial runner and trial.

I think ideally we would move this whole logic just into the trial, which keeps track of it's errors and retry-behavior.

For this, the trial would have to identify the cause of the error (e.g. actor dead, training fn failure, restore failure) and update it's own state accordingly and react in `should_recover()`. Then the trial runner does not have to do any special casing

**Review by krfricke — CHANGES_REQUESTED** [review]
I think we should move this whole logic into the trial to avoid strong coupling between the trial runner and the trial. Basically we want to avoid that the trial runner tells the trial how to update it's internal state.

Instead, we should make the information available to the trial so that it can act based on this. For this case, we would want to know what part of the trial failed.

I can think of a few ways to do that:
1. We introduce `TuneRestoreError`, `TuneTrainingError`, `TuneActorError`, and wrap `exc` in it before passing it to `write_error_log`.
2. We pass something like an `ErrorType` enum

I think I prefer option 1, but open to suggestions.

The trial can then decide which counters to increase and how to write the error log. For instance, it would be helpful to log restore errors as `Restore failure # {}` to distinguish this from training errors in the log file.

For a first step to land it until branch cut, we can just distinguish between restore and non-restore errors, if that's easier.

As a minor refactor note, I think we should rename `write_error_log` to `handle_error` and only call it when an error is actually set (i.e. move the `if exc` into the trial runner and remove the `Optional` annotation in the method).

What do you think?

**Comment by xwjiang2010** [comment]
Ok updated. PTAL!

File: python/ray/tune/experiment/trial.py [hunk 3]
```diff
@@ -638,11 +641,24 @@ def set_experiment_tag(self, experiment_tag):
...
+            exc = exc.exc
+            if self.num_restore_failures >= int(
+                os.environ.get("TUNE_RESTORE_RETRY_NUM", 0)
+            ):
+                # Restore was unsuccessful, try again without checkpoint.
+                self.clear_checkpoint()
+            else:
+                self.num_restore_failures += 1
+                use_restore_counter = True
+        if not use_restore_counter:
```
Comment (krfricke) [inline python/ray/tune/experiment/trial.py:656]: Nit
```suggestion
        if isinstance(exc, _TuneRestoreError):
            exc = exc.exc
            if self.num_restore_failures >= int(
                os.environ.get("TUNE_RESTORE_RETRY_NUM", 0)
            ):
                # Restore was unsuccessful, try again without checkpoint.
                self.clear_checkpoint()
                self.num_failures += 1
            else:
                self.num_restore_failures += 1
        else:
```

**Review by krfricke — APPROVED** [review]
LGTM, minor nit

**Review by Yard1 — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]
approve for docs

File: python/ray/tune/tests/test_tuner_restore.py [hunk 4]
```diff
@@ -437,6 +440,63 @@ def test_tuner_restore_latest_available_checkpoint(
     assert result.metrics["iterations_since_restore"] == 5
 
 
+@pytest.mark.parametrize("retry_num", [0, 1])
+def test_retore_retry(retry_num):
```
Comment (krfricke) [inline python/ray/tune/tests/test_tuner_restore.py:444]:
```suggestion
def test_retore_retry(ray_start_4_cpus, retry_num):
```
should shut down ray gracefully

Comment (xwjiang2010) [reply python/ray/tune/tests/test_tuner_restore.py:444]: yeah I looked at this. Seems like this is a fixture used across board. I wonder why it's running issues now...

**Pull request merged**